### ISSUE TITLE
create/create_key_pair operation in pie

### ISSUE THREAD

**Issue by vbnmmnbv** [OpenKMIP/PyKMIP#265]
Hi Peter,

Maybe I have some misunderstanding of your philosophy of pie, but I think for create/create_key_pair, it's not vey reasonable to only have "name"  and "policy" as the attributes inputs.

 In real scenario, creating a key with other attributes probably is very common. For example, for MAC, I need to set the crypto usage mask for the key in the create() operation.  Or maybe I want to tag a "activate date" with it.  Or a "custom attribute" with it. Of course I can add the attributes later but that's not very neat and also make the operation not atomic.

So do you think it's more reasonable to give them a list of attributes as the input instead of just "name" and "policy". 

I know this will make the user to build all the attributes. But I think that's necessary step to make the pie interface really useful in real scenario.

**Comment by vbnmmnbv** [comment]
If you agree, I am happy to work on it :)

**Comment by PeterHamilton** [comment]
You're not wrong. The original intent for `pie` was to make a simple, obvious interface for KMIP operations. The original `KMIPProxy` requires you to prebuild your attribute objects manually, which is messy and not well documented. It also exposes way too much of the internal library guts, which can change. The plan was always to add on arguments to the `ProxyKmipClient` methods as needed, so adding something like `cryptographic_usage_mask` is fine.

I've toyed with just adding a `**kwargs` field to each client method and calling it a day, but I prefer having explicit arguments. They're self documenting and it just makes things clearer for developers. Of course, with something like `create_key_pair` this could get really messy, since you theoretically could have tons of unique attributes for both public and private keys.

What/how many specific arguments do you need and for which operations?

**Comment by vbnmmnbv** [comment]
Currently in my project, I probably want to add crypto usage mask and "custom attributes" to the create() operation. 

So as a tradeoff, do you think it's a  reasonable design to add an additional argument for create():

`  def create(self, algorithm, length, operation_policy_name=None, name=None, other_attributes=None)`

'other_attributes' is a list and need to be built by the user if he wants to add additional attributes.

Similar to create_key_pair:

```
    def create_key_pair(self,
                        algorithm,
                        length,
                        operation_policy_name=None,
                        public_name=None,
                        private_name=None,
                        other_common_attributes=None,
                        other_public_attributes=None
                        other_private_attributes=None
):
```

**Comment by PeterHamilton** [comment]
Let me look into the `cryptographic_usage_mask`. I think I'm ok with it but there may be some defaults required for certain operations. If we do add it, the argument value should be a list of `enums.CryptographicUsageMask` which the client should internal combine into the mask value.

As for custom attributes, I'm reluctant to add them in to PyKMIP core. By definition a custom attribute needs its own tag(s) and primitive object definitions and it won't be recognized as a standard data type by a vanilla KMIP appliance. Which appliance(s) are you working with? And what custom attributes do you need?

**Comment by PeterHamilton** [comment]
Handling custom attributes is more complicated than that. Both the PyKMIP client and the server (PyKMIP, Thales, etc) will need to know how to process the custom attribute at the byte level, whether it's a data blob, a set of integers, a string, etc. You can't just pass arbitrary values to the client and expect it to send them out over the wire; it needs to know more about each value, its type, and its composition. Your server will need to know how to do comparisons with the attribute value if you want to use it with a `locate` filter.

Is there documentation from Thales on the specification for this custom attribute? It should look like specification details from the KMIP specification. Is that documentation publicly available (I'd be interested in taking a look)?

As for `other_attributes`, it is too open ended for the API. I may be ok with something like `custom_attributes`, since that's allowed in the specification, but again, I need more information.

**Comment by vbnmmnbv** [comment]
Ok. Maybe I have some misunderstanding of custom attributes.  It's not a custom attributes from Thales. It's a custom attribute I want to use for my client application which will interact with a KMIP server(Thales). Let me explain my requirement and maybe you can give me some guidance.

I need to create AES symmetric keys on the KMIP server. They will be of different 'types' from my client's perspective. For example, type A is used for application A and type B is used for application B.
For this purpose, I thought I can either use 'name', 'application specific information' or 'custom attributes'. And I thought 'custom attributes' looks the right way among the three.
So my plan was to create a custom attributes with the name "x-key_type" and for the value, probably use Integer (e.g., 0 stands for type A and 1 stands for type B). Client creates this attribute with the key. Server stores its name and value but doesn't really know the meaning of it.
Later, we client does a locate() passing this custom attribute. All server needs to do is to find a key with the same custom attribute name and value.

Do you think this flow works?  Why the server needs to know anything(except it has a name and integer value)? Or you think I should use 'name' or 'application specific information' instead? Appreciate your inputs.

Thanks,

**Comment by PeterHamilton** [comment]
Ah, ok, this makes a lot more sense. You don't need a `custom_attribute` for this. I would recommend using either a naming scheme with the `name` attribute (e.g., 'Application_A_AES_key_0001') OR take a look at using an `object group`. An `object group` is simply a string label you can apply to objects stored by the KMIP server. Objects can belong to multiple `object group`s and multiple objects can be in the same `object group`. For your use case, you'd simply add `object group` labels of 'Application A' and 'Application B' to each key associated with each application.

PyKMIP already supports object `name`s, so if you went that route you're good to go. We don't fully support `object group`s yet, but I'd be happy to add that in next week sometime. Personally I'd recommend going with `object group`s; it's the closest fit for your use case.

**Comment by vbnmmnbv** [comment]
Thanks for your suggestion. Yeah, I also think 'name' probably would work.
BTW, for my own learning, do you mind letting me know what's the problem of my 'custom attributes' flow as mentioned above? Isn't that also the right way to use custom attributes?
Thanks,

**Comment by PeterHamilton** [comment]
Sure. You could use a `custom attribute` to satisfy your use case; it's just way more involved than what you need (and involves server feature support that is out of your control). You would need to do the following (at a minimum) to create and use a `custom attribute` for PyKMIP:

- Define an official name for it (e.g., 'x-Key-Application')
- Define what type of value you'd need to identify/store the attribute. In your case, it's just a string (an application name or ID). But you could easily add on additional pieces of information associated with the application (e.g., domain, permissions, account information), in which case you'd need to define a new struct to hold all of this information together. Regardless, a new object is needed.
- Define new tags (anything from 540000 - 54FFFF) to uniquely identify any new objects (strings, integers, etc) or structs you define for your attribute. Tags are needed at the byte level when encoding/decoding KMIP requests; without them, the client and server don't know what the values mean nor what they're for. (This has big implications for servers which I'll get to in a second.) (Note too that there are multiple tags for the same type of object. For example, both Issuer (tag 42003B) and Serial Number (tag 420087) are tags for string objects. The way both strings are encoded is the same, but the tag differentiates them. In your case, you'd need at least one new tag for the string or integer value used to differentiate applications.)
- Update the attribute factories to know how to build this new attribute.

This takes care of the client side of things, but you would still have a problem on the server side. KMIP clients and servers must agree on what tags they recognize, otherwise they cannot parse the request/response messages passed between them. If you add a custom attribute to the client, your server has to recognize what it is so that it can store it properly. It's not just a generic key/value store. If the server receives two strings in a request, it uses the tags to parse these values and tell them apart. If you want the server to be able to filter objects by your custom attribute when using the `locate` operation, it has to know how to do comparisons on it and those can change depending on the type. If the server does not recognize your custom attribute/tag, it would likely throw a parse error and reject your request as an invalid message.

The worst case scenario when adding custom attributes is a tag collision, where one brand of KMIP appliance reserves tag 540001 for one type while another brand reserves tag 540001 for a different type. This could create all sorts of problems when storing or retrieving your keys.

Did that help with your understanding? I'm happy to answer more questions if you have any.

Also, the more I think about it, I actually think `application specific information` might be most appropriate for you. It's literally just an `application namespace` string (e.g., 'Application') with a corresponding string value (e.g., 'A', 'B'). It's up to you which one you'd like to use. If `application specific information` seems good, you should check your appliance documentation to see if it supports `application specific information`, specifically if it puts any restraints on the `application namespace` value. Some appliances can be strict with attribute handling. If it looks like supporting your own arbitrary `application specific information` with your appliance could be challenging, just use an `object group` instead.

Regardless, I'll work on making sure all options are available through PyKMIP.

**Comment by vbnmmnbv** [comment]
Thanks a lot for your detailed explanation! Really helpful for me go gain more insights into KMIP.
I see.  So I will evaluate all the options and find the best one.

**Comment by vbnmmnbv** [comment]
Hi,

Previously I said, for the KMIP server I use, "if crypo usage mask is not set, all operations are allowed". I take it back. Actually, just found that's also required.

So I created a pull request #276 trying to handle this to the pie create() operation.

Also sorry for the recent spam I generated (all the pending pull requests). I am now actively working on my project and just need some enhancements to unblock my development.

**Comment by PeterHamilton** [comment]
Not a problem. I'll try to review some of your patches tomorrow. I've been swamped lately.

**Comment by PeterHamilton** [comment]
Hi @vbnmmnbv, I'm going to close this issue for now. I recently added usage mask arguments to the `create_key_pair` method for the `ProxyKmipClient` (see #342). There are still others we discussed that I need to work on, so apologies for the delay in those. Feel free to comment further or file a new issue in the future if you have additional feature requests or info to share. Thanks!
